### PR TITLE
Add /proc/fd link on dup() for files

### PR DESCRIPTION
**Pull request by vtikoo**
Signed-off-by: Vikas Tikoo <vitikoo@microsoft.com>

### CODE REVIEW

File: kernel/fdtable.c [hunk 1]
```diff
@@ -374,6 +374,12 @@ int myst_fdtable_dup(
         new->device = old->device;
         new->object = newobj;
 
+        // add /proc/self/fd/[fd] entry for files
+        if (newfd != oldfd && new->type == MYST_FDTABLE_TYPE_FILE)
+        {
+            myst_add_fd_link(new->device, new->object, newfd);
```
Comment (bodzhang) [inline kernel/fdtable.c:380]: Can this operation fail? If it fails, do we need to clean up anything from fdtable?

Comment (vtikoo) [reply kernel/fdtable.c]: Have updated to closing the new file handle and forward the error.

File: kernel/fdtable.c [hunk 2]
```diff
@@ -374,6 +374,18 @@ int myst_fdtable_dup(
         new->device = old->device;
         new->object = newobj;
 
+        // add /proc/self/fd/[fd] entry for files
+        if (newfd != oldfd && new->type == MYST_FDTABLE_TYPE_FILE)
```
Comment (bodzhang) [inline kernel/fdtable.c:378]: If the oldfd is a pipe, should a link also be added?

Comment (vtikoo) [reply kernel/fdtable.c:378]: Adding /proc/self/fd/[fd] symlink requires the target device(the one which the fd belongs to), to support `fs_realpath`. pipdev.c does not support this currently.

Comment (bodzhang) [reply kernel/fdtable.c:378]: That's unfortunate. The app exposed the issue of missing /dev/fd link created a pipe, dup2 one end of it, and attempted to access it through /dev/fd/[fd].

Comment (vtikoo) [reply kernel/fdtable.c:378]: Is it creating a named pipe? We have an outstanding issue for that - #46.

Comment (bodzhang) [reply kernel/fdtable.c:378]: Not sure whether it's a named pipe. Please add an ATTN comment explaining that once support for /dev/fd/[fd] for other type of descriptors are added, the dup code here need to change.

**Review by bodzhang — APPROVED** [review]

**Pull request merged**